library/common.h

```diff
@@ -68,6 +69,42 @@ extern void (*mbedtls_test_hook_test_fail)( const char * test, int line, const c
  */
 #define MBEDTLS_ALLOW_PRIVATE_ACCESS
 
+/** Return an offset into a buffer.
+ *
+ * This is just the addition of an offset to a pointer, except that this
+ * function also accepts an offset of 0 into a buffer whose pointer is null.
```

Comment:
Err, this accepts any offset into a null buffer, not just `0`. And it might be useful to explain WHY we have this, since otherwise it might get removed at some point in the future when we've forgotten why we ever added it...

Reply:
A null pointer with a nonzero offset is outside the domain of the function. I wondered what to do in this case. There are two reasonable implementations:

* `p == NULL ? NULL : p + n` → null pointer in the misuse case
* `n == 0 ? p : p + n` → invalid addition in the misuse case

I think passing a null pointer through is the least unsafe behavior. On most platforms, a null pointer dereference is a clean crash. Adding an offset to a null pointer might cause a crash on hardened platforms, but on most platforms, it's creating a pointer that might become valid. So I think I picked the safer of the two behaviors.

We could assert, but the function would start to become complicated.

Reply:
> it might be useful to explain WHY we have this

I thought “this function also accepts an offset of 0 into a buffer whose pointer is null” explained why. What else should I say?